### ISSUE TITLE
Update `Punctuated` to handle the `final_value_opt`

### ISSUE THREAD

**Issue by kayagokalp** [FuelLabs/sway#2073]
Currently, if there is no trailing `,` after the last enum variant, the variant is completely removed from the formatted output.

Example:
```
contract;

enum Color {
 Red : (),
 Blue : ()
}
```

is formatted to 

```
contract;

enum Color {
 Red : (),
}
```

**Comment by eureka-cpu** [comment]
This problem extends to `attributes` that don't have a comma like so:
```
#[storage(read)]
```

turns into:
```
#[storage()]
```

I've tested it, and if I add the comma back in:
```
#[storage(read,)]
```
it returns back the same line...

this may actually be a problem with `sway-parse`!

**Comment by eureka-cpu** [comment]
cc @canndrew for thoughts/suggestions

**Comment by eureka-cpu** [comment]
So when you get the inner scope of a punctuated pair there isn't an optional punctuation, it's as if it assumes there will always be punctuation. This is my interpretation since the information we are looking at from the parser is `Punctuated<T, P>` and not `Punctuated<T, Option<P>>`. @canndrew correct me if I'm mistaken here 😅

**Comment by XFFXFF** [comment]
I investigated it and we may need to take `final_value_opt` into consider when we try to get the fields of the enum during formatting, now we only consider `value_separator_pairs`
https://github.com/FuelLabs/sway/blob/19586172e455a499d09d03b0aaca3d44bb47665c/sway-parse/src/punctuated.rs#L3-L7
https://github.com/FuelLabs/sway/blob/19586172e455a499d09d03b0aaca3d44bb47665c/sway-fmt-v2/src/items/item_enum.rs#L32

I printed the `ItemEnum` that parser gave us, there is a filed called `Blue` in `final_value_opt`
```
ItemEnum {
    visibility: None,
    enum_token: ...,
    name: ...,
    generics: None,
    where_clause_opt: None,
    fields: Braces {
        inner: Punctuated {
            value_separator_pairs: [
                (
                    TypeField {
                        name: Ident {
                            name_override_opt: None,
                            span: Span {
                                src (ptr): 0x00007fcc90000c10,
                                path: None,
                                start: 25,
                                end: 28,
                                as_str(): "Red",
                            },
                        },
                        colon_token: ColonToken {
                            span: Span {
                                src (ptr): 0x00007fcc90000c10,
                                path: None,
                                start: 29,
                                end: 30,
                                as_str(): ":",
                            },
                        },
                        ty: Tuple(
                            Parens {
                                inner: Nil,
                                span: Span {
                                    src (ptr): 0x00007fcc90000c10,
                                    path: None,
                                    start: 31,
                                    end: 33,
                                    as_str(): "()",
                                },
                            },
                        ),
                    },
                    CommaToken {
                        span: Span {
                            src (ptr): 0x00007fcc90000c10,
                            path: None,
                            start: 33,
                            end: 34,
                            as_str(): ",",
                        },
                    },
                ),
            ],
            final_value_opt: Some(
                TypeField {
                    name: Ident {
                        name_override_opt: None,
                        span: Span {
                            src (ptr): 0x00007fcc90000c10,
                            path: None,
                            start: 36,
                            end: 40,
                            as_str(): "Blue",
                        },
                    },
                    colon_token: ColonToken {
                        span: Span {
                            src (ptr): 0x00007fcc90000c10,
                            path: None,
                            start: 41,
                            end: 42,
                            as_str(): ":",
                        },
                    },
                    ty: Tuple(
                        Parens {
                            inner: Nil,
                            span: Span {
                                src (ptr): 0x00007fcc90000c10,
                                path: None,
                                start: 43,
                                end: 45,
                                as_str(): "()",
                            },
                        },
                    ),
                },
            ),
        },
        span: ...,
    },
}
```

**Comment by eureka-cpu** [comment]
Hey @XFFXFF thanks for investigating, I did take into account the `final_value_opt` in my `generics` PR here: #2110 but I wasn't sure how to go about getting any formattable information from its fields

**Comment by eureka-cpu** [comment]
Ah, I see now. I'll take another whack at it, thanks again!

**Comment by eureka-cpu** [comment]
To put it in other words @mohammadfawaz & @kayagokalp the `final_value_opt` is the option to have a value with no `CommaToken` at the end. Since `enum` wasn't taking that into account, it wasn't getting carried over after formatting. This will be closed when #2110 gets merged 🥳
